adaptris/interlok-csv
# Add CSV Schema validator

## Motivation

Add support for CSV Schema using the reference implementation https://github.com/digital-preservation/csv-validator

## Modification

- Made interlok-csv a multi-module
- Added `interlok-csv-schema` as a new optional component
    - since this is has a large number of dependencies on scala, it's a new optional component

## PR Checklist

- [x] Added supporting annotations (like @XStreamAlias / @ComponentProfile)
- [x] Added DefaultValue annotation when there is a default value (e.g. @DefaultValue('true'))
- [x] Added validation annotation (@NotNull...) when required and add @Valid when nested class has some validation
- [x] Checked that @NotNull and @NotBlank annotations have a meaningful message when appropriate.
- [x] Checked that new 3rd party dependencies are appropriately licensed
- [x] Added comments explaining the "why" and the intent of the code wherever it would not be obvious for an unfamiliar reader
- [x] Tested new/updated components in the UI and at runtime in an Interlok instance
- [x] Reviewed java class members so that missing annotations are added (InputFieldDefault/ComponentProfile etc)
- [x] Checked that javadoc generation doesn't report errors
- [x] Checked the display of the component in the UI

## Result

No change for the user unless they add csv-schema as an optional dependency

## Testing

- Standard interlok parent gradle; including the base line validator as dependencies; include interlok-csv as well.
```
  interlokRuntime ("uk.gov.nationalarchives:csv-validator-java-api:1.1.5")
  interlokRuntime("com.gilt:gfc-semver_2.11") {
    version {
      strictly '0.0.5'
    }
  }
```
- Build the jar and copy it into the right place.
- Take the CSV schema file from the unit tests (`warnings.csvs`)
- You can use the configuration :
```xml
<validate-csv-against-schema>
  <schema-file>/path/to/warnings.csvs</schema-file>
</validate-csv-against-schema>
```
- Choose your input from _passes.csv|warnings.csv|warningsAndErrors.csv_ and check your workflow behaviour.
